isotovideo

```diff
@@ -109,73 +108,71 @@ chdir $options{workdir} if $options{workdir};
+my $runner;
 
 # make sure all commands coming from the backend will not be in the
 # developers's locale - but a defined english one. This is SUSE's
 # default locale
 $ENV{LC_ALL} = 'en_US.UTF-8';
 $ENV{LANG} = 'en_US.UTF-8';
 
-diag(_get_version_string());
+eval {
+    diag(_get_version_string());
 
-# enable debug default when started from a tty
-$log::direct_output = $options{debug};
+    # enable debug default when started from a tty
+    $log::direct_output = $options{debug};
 
-$bmwqemu::scriptdir = $installprefix;
+    $bmwqemu::scriptdir = $installprefix;
 
-my $runner = OpenQA::Isotovideo::Runner->new;
-$runner->_init_bmwqemu(@ARGV);
-$runner->prepare;
-$runner->start_autotest;
-$runner->create_backend;
+    $runner = OpenQA::Isotovideo::Runner->new;
+    $runner->_init_bmwqemu(@ARGV);
+    $runner->prepare;
+    $runner->start_autotest;
+    $runner->create_backend;
 
-spawn_debuggers;
+    spawn_debuggers;
 
-$runner->handle_commands;
+    $runner->handle_commands;
 
-$return_code = 0;
+    $return_code = 0;
 
-# enter the main loop: process messages from autotest, command server and backend
-$runner->run;
+    # enter the main loop: process messages from autotest, command server and backend
+    $runner->run;
 
-# terminate/kill the command server and let it inform its websocket clients before
-$runner->stop_commands('test execution ended');
+    # terminate/kill the command server and let it inform its websocket clients before
+    $runner->stop_commands('test execution ended');
 
-if ($runner->testfd) {
-    # unusual shutdown
-    $return_code = 1;    # uncoverable statement
-    CORE::close $runner->testfd;    # uncoverable statement
-    $runner->stop_autotest();    # uncoverable statement
-}
-
-diag 'isotovideo ' . ($return_code ? 'failed' : 'done');
-
-my $clean_shutdown;
-if (!$return_code) {
-    eval {
-        $clean_shutdown = $bmwqemu::backend->_send_json({cmd => 'is_shutdown'});
-        diag('backend shutdown state: ' . ($clean_shutdown // '?'));
-    };
-
-    # don't rely on the backend in a sane state if we failed - just stop it later
-    eval { bmwqemu::stop_vm(); };
-    if ($@) {
-        bmwqemu::serialize_state(component => 'backend', msg => "unable to stop VM: $@", error => 1);
-        $return_code = 1;
+    if ($runner->testfd) {
+        # unusual shutdown
+        $return_code = 1;    # uncoverable statement
+        CORE::close $runner->testfd;    # uncoverable statement
+        $runner->stop_autotest();    # uncoverable statement
     }
-}
 
-# read calculated variables from backend and tests
-bmwqemu::load_vars();
+    diag 'isotovideo ' . ($return_code ? 'failed' : 'done');
+
+    my $clean_shutdown;
+    if (!$return_code) {
+        eval {
+            $clean_shutdown = $bmwqemu::backend->_send_json({cmd => 'is_shutdown'});
+            diag('backend shutdown state: ' . ($clean_shutdown // '?'));
+        };
+
+        # don't rely on the backend in a sane state if we failed - just stop it later
```

Comment:
I would add a `local $@` here, otherwise it might survive the outer eval.
(but I'm currently looking at the diff from my phone)

Reply:
Not sure why that would be the case. Nested `eval`s seem to work as expected:

```
  DB<6> eval { print "foo\n"; eval { die "bar\n"; }; if($@) { print "inner: $@" } }; if($@) { print "outer: $@" }
foo
inner: bar

  DB<7> eval { die "foo\n"; eval { die "bar\n"; }; if($@) { print "inner: $@" } }; if($@) { print "outer: $@" }
outer: foo

  DB<8> eval { print "foo\n"; eval { die "bar\n"; }; if($@) { print "inner: $@" } die "baz\n" }; if($@) { print "outer: $@" }
foo
inner: bar
outer: baz
```

Reply:
Tina might be thinking of a misfeature in older Perl versions. We now depend on Perl 5.20 though, so this shouldn't be a concern here.

Reply:
ah ok, thanks